### ISSUE TITLE
abstract out mpl plotting layer

### ISSUE THREAD

**Issue by nikhartman** [NSLS-II/bluesky#1174]
Looking at this package and similar work across github, it seems you all have done a nice job of making live plotting work reliably. 

I am interested in trying to abstract some of that out to apply it in different contexts. Here is a quick sketch of a simple experiment (mock) experiment loop, in a notebook cell, with a few comments of what, I think, needs to happen. Hoping you all can point me in the right direction. 

```
%matplotlib notebook

import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np

# kicker required?

xarr = np.linspace(-2*np.pi, 2*np.pi, 200)
yarr = np.full(len(xarr), np.nan)

# initialize LivePlot?

def setx(xval):
    pass

def gety(xval):
    return np.sin(xval)

for i, x in enumerate(x):
    setx(x)
    time.sleep(0.02) # not actually waiting here
    yarr[i] = gety(x)
    # trigger LivePlot update
```

**Comment by danielballan** [comment]
We think of `LivePlot` and friends as a translation from our [event model](https://nsls-ii.github.io/event-model/data-model.html) to matplotlib Artists. The event model is just dictionaries laid out a certain way so that data consumers like LivePlot know where to go to find the information they need to---for example---assign axis labels with units. If all you have is one array, `line = plt.plot(...); while ...: line.setdata(...)` is the simplest way to go. But if you might have _two_ arrays that need to be labeled in a legend, or units that you want to display, etc., buying into the event model gives you something.

```py
from bluesky.callbacks.mpl_plotting import LivePlot                                                                                                                                           
import event_model                                                                                                                                                                            
import time                                                                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                              
def mock_data_source():                                                                                                                                                                       
    for i in range(10):                                                                                                                                                                       
        yield i, i**2                                                                                                                                                                         
                                                                                                                                                                                              
# Describe your data.                                                                                                                                                                         
# You might have more than two fields here, including nonscalar fields with some shape.
# You may optionally add a 'units' key to the dicts inside here.                                                                                                       
data_keys= {                                                                                                                                                                                  
    'temp': dict(source='temperature controller', dtype='number', shape=[]),                                                                                                                  
    'I':  dict(source='ammeter', dtype='number', shape=[])}                                                                                                                                   
# What do you want to plot specifically for this LivePlot? You might set up a couple on the same axes.                                                                                        
lp = LivePlot('I', 'temp')  # takes optional ax=... arg                                                                                                                                       
                                                                                                                                                                                              
# Boilerplate that is handled by the bluesky RunEngine in our case...                                                                                                                         
# I can imagine ways to streamline this for users who aren't using the RunEngine.                                                                                                             
run_bundle = event_model.compose_run()                                                                                                                                                        
lp('start', run_bundle.start_doc)                                                                                                                                                             
stream_bundle = run_bundle.compose_descriptor(                                                                                                                                                
    name='primary',  # arbitrary label for this logical table ("stream") of Events                                                                                                                                                                          
    data_keys=data_keys)                                                                                                                                                                      
lp('descriptor', stream_bundle.descriptor_doc)                                                                                                                                                
                                                                                                                                                                                              
# And now the update loop:                                                                                                                                                                    
for temp, current in mock_data_source():                                                                                                                                                      
    # We put hardware timestamps here, but if you don't have them you can just use the current time.                                                                                          
    now = time.time()                                                                                                                                                                         
    event = stream_bundle.compose_event(                                                                                                                                                      
        data={'temp': temp, 'I': current},                                                                                                                                                    
        timestamps={'temp': now, 'I': now})                                                                                                                                                   
    lp('event', event)                                                                                                                                                                        
                                                                                                                                                                                              
# For completeness:                                                                                                                                                                           
lp('stop', run_bundle.compose_stop())  # doesn't matter for LivePlot, but matters for some things
```

As written, this would run in a blocking fashion until `mock_data_source` is exhausted. If you are looking to run this in the background on a thread, that's where the kicker etc. get involved.

**Comment by nikhartman** [comment]
Ok. I follow that. What you’ve done in this quick example is a nice wrapper around updating a mpl figure one point at a time, while keeping track of data/metadata in a slick way.

What I am aiming for is an implementation of updating plots in the background. Can we work up an example without using the full Run Engine? You could convince me that is the thing to do. But, I’m interested in exploring this idea first.

**Comment by danielballan** [comment]
Reasonable to explore. I'm not completely clear on where the blocking calls are in your sketch. Are `setx` and `gety` reading a database or extracting data from messages that originate from the hardware? Does `setx` move a setpoint and block until it is reached?

**Comment by nikhartman** [comment]
`getx()` and `setx()` block for at least 20 ms. In a real experiment, it would be a collection of measurements or setpoints run in parallel. The `time.sleep()` call is a settling time that is set by properties of the system being measured. It is usually 20 - 100ms. That waiting time is usually the factor that limits measurement times. It would be silly to wait for plot updates at the end of the loop when there is a potential 100ms of wasted time coming up on the next iteration.

**Comment by nikhartman** [comment]
@danielballan Pretty sure the answer is this simple:

```python
install_nb_kicker()

class LiveLine():
    
    ''' a class to handle initializing and updating live line plots '''
    
    def __init__(self, xdata, ax, **kwargs):
        # assume the full x array is known at the start
        # fill array for y data and add y data as it comes in
        
        self.x = xdata
        self.y = np.full(len(self.x), np.nan)
        self.ax = ax
        
        self.line, = self.ax.plot(self.x, self.y)
        self.ax.set_xlim(self.x.min(), self.x.max())
    
    def add_ydata(self, idx, ydata):
        # add y data
        self.y[idx] = ydata
    
    def update_plot(self):
        self.line.set_ydata(self.y)
        self.ax.relim(visible_only=True)
        self.ax.autoscale_view(tight=True, scalex=False, scaley=True)
        self.ax.figure.canvas.draw_idle()

def setx(xval):
    pass

def gety(xval):
    return np.sin(xval)

n=100
x_array = np.linspace(-2*np.pi, 2*np.pi, n)

fig, ax = plt.subplots(1,1)
ll = LiveLine(x_array, ax)

tstart = time.time()
for i, x in enumerate(x_array):

    setx(x) # go to setpoint

    await asyncio.sleep(0.02) # wait settling time
    
    ll.add_ydata(i, gety(x)) # append new y data
    ll.update_plot()
    
print(1000*(time.time()-tstart)/n)
```

**Comment by stan-dot** [comment]
@danielballan is this out of scope? or maybe a good first issue?
